### DISCUSSION TITLE
Error when querying over longer time periods

### DISCUSSION THREAD

**Discussion by sptrakesh** [questdb/questdb#5185] [Ideas]
Hello,

We are frequently running into errors trying to retrieve aggregated information over longer time periods (more than 3 months for instance).  The data set is not very large, generally under 1m rows in table (usually a lot lower).  We very rarely perform out-of-order writes.  Usual error looks like the following:

```Status: 500. Message: db query error: pq: could not open read-only [file=/var/lib/questdb/db/reelHistory~55/_cv]```

These error increase in frequency the longer the database has been up.  Restart definitely helps as a short term fix.

We have not experimented much with configuration parameters, so wondering if this is just a case of us needing to tweak the settings.   Running via docker, the following environment variables are specified:

```
    -e QDB_PG_READONLY_USER_ENABLED=true \
    -e QDB_PG_SELECT_CACHE_ENABLED=true \
    -e QDB_PG_NET_CONNECTION_LIMIT=128 \
    -e QDB_CAIRO_SNAPSHOT_INSTANCE_ID=98cf4d56-d759-440a-be03-edea3e4cb47d \
    -e QDB_CAIRO_WRITER_DATA_APPEND_PAGE_SIZE=1M \
    -e QDB_CAIRO_O3_COLUMN_MEMORY_SIZE=1M \
```

We are not specifying any memory parameters when running the container.  Generally the container consumes about 1G of RAM, but we have more that could be pre-allocated if that helps.  Thanks for any suggestions on working around this issue.

Rakesh

**Comment by puzpuzpuz** [comment]
@sptrakesh could you check if you have increased kernel defaults for max open files and mmap areas? Very likely that you're hitting this limits.
https://questdb.io/docs/deployment/capacity-planning/#os-configuration

**Reply by sptrakesh** [reply]
Hi @puzpuzpuz we have system wide settings modified via `/etc/sysctl.conf` as on that page:

```
fs.file-max=1048576
vm.max_map_count=1048576
```

We were unable to increase the user limits are documented on that page, tried again and get the same error:

```ulimit: 1048576: limit exceeded [Operation not permitted]```

**Reply by puzpuzpuz** [reply]
The server checks the limits when starting and prints them. If the limits are higher than the defaults, you should see something like this in the logs:
```
2024-11-22T13:08:57.104659Z A server-main fs.file-max checked [limit=1048576]
2024-11-22T13:08:57.104760Z A server-main vm.max_map_count checked [limit=1048576]
```

Otherwise, there will be warnings printed.

Could you check your server logs to understand if your limit modifications were applied?

**Reply by sptrakesh** [reply]
Looks like only one of the settings applies (possibly due to per user limits not being increased):

```
2024-11-22T13:15:58.567979Z A server-main fs.file-max limit is too low [limit=65536] (SYSTEM COULD BE UNSTABLE)
2024-11-22T13:15:58.568348Z A server-main vm.max_map_count checked [limit=1048576]
2024-11-22T13:15:58.568489Z A server-main make sure to increase fs.file-max and vm.max_map_count limits:
https://questdb.io/docs/deployment/capacity-planning/#os-configuration
```

I have found instructions for updating the per user limits (setting in `/etc/security/limits.conf` and `/etc/security/limits.d/questdbdb` files).  I will see what difference this makes.

**Reply by sptrakesh** [reply]
Setting the per user limits does not seem to resolve the issue with `fs.file-max`.  I tried restarting the docker daemon and finally rebooted the EC2 host itself.

```shell
# on EC2 host
$ ulimit -Sn                                                     
1048576
```

```
# on questdb container
# ulimit -Sn
32768
# ulimit -Hn
65536
```

I see the container has its own `/etc/security/limits.conf` and `/etc/security/limits.d/` directories.  Should I mount these files from the host machine to over-write these local files on the container?

**Reply by sptrakesh** [reply]
Had to specify `--ulimit nofile=1048576:1048576` in the `docker run` command to pass the requirements to the container.  The warnings are now gone.  Will see how this affects runtime capabilities.

**Reply by puzpuzpuz** [reply]
Thanks for the update! The increased limit should fix your original issue. In the meantime, we'll add OS error code to the errors returned by PGWire interface, so that it's easier to diagnose such issues.
